### PR TITLE
feat: support for running feature files

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
This PR adds support for running feature files, without changing how the TAGS filters are used.

- extracts common logic to run the tests to a common-test.mk file.
- create a Makefile for each test suite, providing each test suite the run tests command, like in a build system for a multi-module project. This new file imports the new common file.
- adds support to filter the test runner by feature file, so that the run tests command is able to select one or more feature files.

We are also documenting how the CI is spreading the scenarios in the workers, using the `.e2e-tests.yaml` descriptor.

Finally, it improves the docs related the variables that affect the build, including TAGS and FEATURES.

## Why is it important?
This PR is the first step to simplify how the scenarios are spread in the parallel workers: nowadays it's using TAGS, but and the end of the day we are using feature files (using the tag for the entire feature file).

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have run the Unit tests (`make unit-test`), and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [ ] I have noticed new Go dependencies (run `make notice` in the proper directory)

## How to test this PR locally

**Default execution:**
It will run all feature files, filtered by the `backend_processes && debian && install` tags.
```shell
TAGS="backend_processes && debian && install" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE BEATS_USE_CI_SNAPSHOT=true  DEVELOPER_MODE=true make -C e2e         functional-test
```

**Running with features:**
It will run the `fleet_mode_agent.feature` file, filtering by the `backend_processes && debian && install` tags.

```shell
FEATURES="fleet_mode_agent.feature" TAGS="backend_processes && debian && install" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE BEATS_USE_CI_SNAPSHOT=true  DEVELOPER_MODE=true make -C e2e functional-test
```

## Related issues

- Relates #1191
- Relates #373 (as after this PR we will be able to filter by feature file instead of the parent tag, being possible to refactor child tags without changing execution)

## Follow-ups
We need to update the e2e YAML descriptor to match this new features approach, updating the Jenkinsfile to spread feature files into the parallel execution. Besides that, we could generate the descriptor on the fly, based on the existing feature files.

### CODE REVIEW

**Comment by mdelapenya** [comment]
jenkins run the tests please

File: e2e/README.md [hunk 1]
```diff
@@ -82,8 +82,13 @@ The anatomy of a feature file is:
...
+
+Adding a new feature file will require to check [the aforementioned descriptor file](../.ci/.e2e-tests.yaml). If the tags in the new file are not there, you should add a new parallel branch under the main test suite, or update the tags to add the new scenarios in an existing parallel branch.
+
 ### Configuration files
-It's possible that there will exist configuration YAML files in the test suire. We recommend locating them under the `configurations` folder in the suite directory. The name of the file will represent the feature to be tested (i.e. `apache.yml`). In this file we will add those configurations that are exclusive to the feature to be tests.
+It's possible that there will exist configuration YAML files in the test suire. We recommend locating them under the `configurations` folder under the config directory for compose files (both profiles and services). The name of the file will represent the feature to be tested (i.e. `kibana.config.yml`). In this file we will add those configurations that are exclusive to the feature to be tests.
```
Comment (mdelapenya) [inline e2e/README.md:91]:
```suggestion
It's possible that there will exist configuration YAML files in the test suite. We recommend locating them under the `configurations` folder under the config directory for compose files (both profiles and services). The name of the file will represent the feature to be tested (i.e. `kibana.config.yml`). In this file we will add those configurations that are exclusive to the feature to be tested.
```

File: e2e/README.md [hunk 2]
```diff
@@ -82,8 +82,13 @@ The anatomy of a feature file is:
     - **But**: Used within any of the above clauses, it must tell an ocational reader a secondary preparation (Given), trigger (When), or output (Then) that must not be present.
 - **Examples:**: this `markdown table` will represent the elements to interpolate in the existing dynamic variables in the use case, being each column header the name of the different variables in the table. Besides that, each row will result in a test execution.
 
+#### Feature files and the CI
+There is [a descriptor file for the CI](../.ci/.e2e-tests.yaml) in which we define the parallel branches that will be created in the execution of a job.This YAML file defines suites and tags. A suite represents each test suite directory under the `e2e/_suites` directory, and the tags represent the tags will be passed to the test runner to filter the test execution. Another configuration we define in this file is related to the capabilities to run certain tags at the pull request stage, using the `pullRequestFilter` child element. This element will be appended to the tags used to filter the test runner.
```
Comment (cachedout) [inline e2e/README.md:86]:
```suggestion
There is [a descriptor file for the CI](../.ci/.e2e-tests.yaml) in which we define the parallel branches that will be created in the execution of a job. This YAML file defines suites and tags. A suite represents each test suite directory under the `e2e/_suites` directory, and the tags represent the tags will be passed to the test runner to filter the test execution. Another configuration we define in this file is related to the capabilities to run certain tags at the pull request stage, using the `pullRequestFilter` child element. This element will be appended to the tags used to filter the test runner.
```

File: e2e/README.md [hunk 3]
```diff
@@ -82,8 +82,13 @@ The anatomy of a feature file is:
...
+
+Adding a new feature file will require to check [the aforementioned descriptor file](../.ci/.e2e-tests.yaml). If the tags in the new file are not there, you should add a new parallel branch under the main test suite, or update the tags to add the new scenarios in an existing parallel branch.
+
 ### Configuration files
-It's possible that there will exist configuration YAML files in the test suire. We recommend locating them under the `configurations` folder in the suite directory. The name of the file will represent the feature to be tested (i.e. `apache.yml`). In this file we will add those configurations that are exclusive to the feature to be tests.
+It's possible that there will exist configuration YAML files in the test suite. We recommend locating them under the `configurations` folder under the config directory for compose files (both profiles and services). The name of the file will represent the feature to be tested (i.e. `kibana.config.yml`). In this file we will add those configurations that are exclusive to the feature to be tested.
```
Comment (cachedout) [inline e2e/README.md:91]:
```suggestion
It's possible that there configuration YAML files will exist in the test suite. We recommend locating them under the `configurations` folder under the config directory for compose files (both profiles and services). The name of the file will represent the feature to be tested (i.e. `kibana.config.yml`). In this file we will add those configurations that are exclusive to the feature to be tested.
```

File: e2e/commons-test.mk [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ifeq ($(ELASTIC_APM_ACTIVE),true)
+ifeq ($(ELASTIC_APM_ENVIRONMENT),ci)
+JENKINS_STATS_SECRET?=secret/observability-team/ci/jenkins-stats
+export APM_SECRET_TOKEN?=$(shell vault read -field apmServerToken "$(JENKINS_STATS_SECRET)")
+export APM_SERVER_URL?=$(shell vault read -field apmServerUrl "$(JENKINS_STATS_SECRET)")
+endif
+
+ifeq ($(ELASTIC_APM_ENVIRONMENT),local)
+export APM_SECRET_TOKEN?=
+export APM_SERVER_URL?=http://localhost:8200
```
Comment (cachedout) [inline e2e/commons-test.mk:24]: There are two colons here. Should this value be quoted?

File: e2e/README.md (hunk 3 above)
Comment (mdelapenya) [reply e2e/README.md:91]:
```suggestion
It's possible that the configuration YAML files will exist in the test suite. We recommend locating them under the `configurations` folder under the config directory for compose files (both profiles and services). The name of the file will represent the feature to be tested (i.e. `kibana.config.yml`). In this file we will add those configurations that are exclusive to the feature to be tested.
```

File: e2e/commons-test.mk (hunk 4 above)
Comment (mdelapenya) [reply e2e/commons-test.mk:24]: I noticed that Makefiles behaves differently than shell scripts, and treats double quotes as part of the string 🤷‍♀️ 

>$ SUITE="fleet" TAGS="fleet_mode_agent && nightly" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE BEATS_USE_CI_SNAPSHOT=false DEVELOPER_MODE=true ELASTIC_APM_ENVIRONMENT=local ELASTIC_APM_ACTIVE=true make -C e2e functional-test 
GOARCH=amd64 go get -v github.com/cucumber/godog/cmd/godog@v0.11.0
cd _suites/fleet && \
	OP_LOG_LEVEL=TRACE \
	OP_LOG_INCLUDE_TIMESTAMP=TRUE \
	TIMEOUT_FACTOR=3 \
	STACK_VERSION= \
	DEVELOPER_MODE=true \
	ELASTIC_APM_SERVICE_NAME="E2E Tests" \
	ELASTIC_APM_CENTRAL_CONFIG="false" \
	ELASTIC_APM_GLOBAL_LABELS= \
	ELASTIC_APM_SERVICE_VERSION="" \
	ELASTIC_APM_ENVIRONMENT="local" \
	ELASTIC_APM_SECRET_TOKEN="" \
	ELASTIC_APM_SERVER_URL="http://localhost:8200" \
	godog --format=pretty --tags "fleet_mode_agent && nightly && ~skip"
^Cmake: *** [functional-test] Interrupt: 2

Comment (adam-stokes) [reply e2e/commons-test.mk:24]: I think Make is special in this case in that it will interpret both `'` and `"` as literal

**Review by adam-stokes — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Review by kuisathaverat — APPROVED** [review]

**Pull request merged**